src/java.base/share/classes/java/nio/Direct-X-Buffer.java.template

```diff
@@ -78,18 +79,42 @@ class Direct$Type$Buffer$RW$$BO$
         }
 
         public void run() {
-            UNSAFE.freeMemory(address);
-            Bits.unreserveMemory(size, capacity);
+            try {
+                UNSAFE.freeMemory(address);
+                Bits.unreserveMemory(size, capacity);
+            } catch (Throwable x) {
+                // Old jdk.internal.ref.Cleaner behavior: when it fails, VM exits.
+                if (System.err != null) {
+                    new Error("Cleaner terminated abnormally", x).printStackTrace();
+                }
+                System.exit(1);
+            }
         }
     }
 
-    private final Cleaner cleaner;
+    // Singleton Cleaner for all relevant Buffer instances.
+    // Lazily initialized to avoid bootstrapping circularities involving <clinit>.
+    static class CleanerHolder {
+        private static final Cleaner CLEANER = Cleaner.create();
 
-    public Cleaner cleaner() { return cleaner; }
+        public static Cleanable register(Object buffer, Runnable action) {
```

Comment:
Why is this public?

Reply:
It is a public method in a private nested class. This method is used from the enclosing class, so private would be weird? I can drop public, though, let it be package-private. Done in new commit.